File: src/main/java/se/landsmann/sufusku/Matrix.java
package se.landsmann.sufusku;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Matrix {

    private List<List<Cell>> data = new ArrayList<>(9);

    {
        for (int i = 0; i < 9; i++) {
            List<Cell> row = new ArrayList<>(9);
            data.add(row);
            for (int j = 0; j < 9; j++) {
                row.add(new Cell(i, j, getGroupIndex(i, j)));
            }
        }

        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                Cell cell = getCell(row, col);
                cell.setRow(getRow(row));
                cell.setCol(getCol(col));
                cell.setGroup(getGroup(row, col));
            }
        }
    }

    private List<Cell> getGroup(int row, int col) {

        GroupBounderies gb = GroupBounderies.getGroupBounderies(row, col);

        List<Cell> group = new ArrayList<>(9);

        for (int i = gb.minX; i < gb.maxX; i++) {
            for (int j = gb.minY; j < gb.maxY; j++) {
                group.add(getCell(i, j));
            }
        }
        return group;
    }

    private int getGroupIndex(int row, int col) {

        GroupBounderies gb = GroupBounderies.getGroupBounderies(row, col);

        if (gb.minX == 0) {
            if (gb.minY == 0) {
                return 1;
            } else if (gb.minY == 3) {
                return 2;
            } else {
                return 3;
            }
        } else if (gb.minX == 3) {
            if (gb.minY == 0) {
                return 4;
            } else if (gb.minY == 3) {
                return 5;
            } else {
                return 6;
            }
        } else {
            if (gb.minY == 0) {
                return 7;
            } else if (gb.minY == 3) {
                return 8;
            } else {
                return 9;
            }
        }
    }

    public Cell getCell(int row, int col) {
        return data.get(row).get(col);
    }

    public List<Cell> getRow(int i) {
        return data.get(i);
    }

    public List<Cell> getCol(int i) {
        return data.stream()
                .map(row -> row.get(i))
                .collect(Collectors.toList());
    }

    public int getCellValue(int row, int col) {
        return getCell(row, col).getValue();
    }

    public void setCellValue(int row, int col, int value) {
        Cell cell = getCell(row, col);
        cell.setValue(value);
        checkMatrixForCraziness();
    }

    private void checkMatrixForCraziness() {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                boolean isCrazy = checkIfCellIsCrazy(i, j);
                getCell(i, j).setIsCrazy(isCrazy);
            }
        }
    }

    private boolean checkIfCellIsCrazy(int row, int col) {
        int value = getCellValue(row, col);

        boolean isCrazy = getRow(row).stream()
                .filter(c -> c.getValue() != 0 && c.getColIndex() != col)
                .anyMatch(c -> c.getValue() == value);

        isCrazy |= getCol(col).stream()
                .filter(c -> c.getValue() != 0 && c.getRowIndex() != row)
                .anyMatch(c -> c.getValue() == value);

        isCrazy |= getGroup(row, col).stream()
                .filter(c -> c.getValue() != 0 && c.getRowIndex() != row && c.getColIndex() != col)
                .anyMatch(c -> c.getValue() == value);

        return isCrazy;
    }

    public String toJson() {

        JSONArray rows = new JSONArray();

        for (int i = 0; i < 9; i++) {

            JSONArray row = new JSONArray();
            rows.put(row);

            for (int j = 0; j < 9; j++) {
                JSONObject cell = new JSONObject();
                Cell c = getCell(i, j);
                cell.put("value", c.getValue());
                cell.put("numbers", c.getNumbers());
                cell.put("isCrazy", c.isCrazy());
                row.put(cell);
            }
        }

        JSONObject json = new JSONObject();
        json.put("matrix", rows);

        return json.toString();
    }

    public void reset() {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                getCell(i, j).reset();
            }
        }
    }
}

File: src/matrix.js
var Matrix = function () {

    this.matrix = new Array(9);

    for (row = 0; row < 9; row++) {
        this.matrix[row] = new Array(9);
        for (cell = 0; cell < 9; cell++) {
            this.matrix[row][cell] = new Cell(row, cell, getGroupIndex(row, cell));
        }
    }

    function getGroup(matrix, row, col) {

        var gb = getGroupBounderies(row, col);

        var group = [];

        for (var i = gb.minX; i < gb.maxX; i++) {
            for (var j = gb.minY; j < gb.maxY; j++) {
                group.push(matrix[i][j]);
            }
        }
        return group;
    }

    function getGroupIndex(row, col) {

        var gb = getGroupBounderies(row, col);

        if (gb.minX === 0) {
            if (gb.minY === 0) {
                return 1;
            } else if (gb.minY === 3) {
                return 2;
            } else {
                return 3;
            }
        } else if (gb.minX === 3) {
            if (gb.minY === 0) {
                return 4;
            } else if (gb.minY === 3) {
                return 5;
            } else {
                return 6;
            }
        } else {
            if (gb.minY === 0) {
                return 7;
            } else if (gb.minY === 3) {
                return 8;
            } else {
                return 9;
            }
        }
    }


    function getGroupBounderies(row, col) {

        var minX, maxX, minY, maxY;

        if (row >= 0 && row < 3) {
            minX = 0;
            maxX = 3;
        } else if (row >= 3 && row < 6) {
            minX = 3;
            maxX = 6;
        } else {
            minX = 6;
            maxX = 9;
        }

        if (col >= 0 && col < 3) {
            minY = 0;
            maxY = 3;
        } else if (col >= 3 && col < 6) {
            minY = 3;
            maxY = 6;
        } else {
            minY = 6;
            maxY = 9;
        }

        return {minX: minX, maxX: maxX, minY: minY, maxY: maxY}
    }

    var removeNumberFromCell = function (cell, value) {
        cell.numbers = cell.numbers.replace(new RegExp(value), '');
    };


    function checkIfCellIsCrazy(matrix, row, col) {
        var value = matrix[row][col].value;
        var isCrazy = false;
        var i, cell;

        if (value === '') {
            return false;
        }

        for (i = 0; i < 9; i++) {
            cell = matrix[row][i];
            if (cell.col !== col && cell.value === value) {
                isCrazy = true;
            }
        }

        for (i = 0; i < 9; i++) {
            cell = matrix[i][col];
            if (cell.row !== row && cell.value === value) {
                isCrazy = true;
            }
        }

        var group = getGroup(matrix, row, col);

        for (i = 0; i < 9; i++) {
            cell = group[i];
            if (cell.row !== row && cell.col !== col && cell.value === value) {
                isCrazy = true;
            }
        }

        return isCrazy;
    }

    this.setValue = function (row, col, value) {

        var cell;
        for (cell = 0; cell < 9; cell++) {
            removeNumberFromCell(this.matrix[row][cell], value);
        }

        for (cell = 0; cell < 9; cell++) {
            removeNumberFromCell(this.matrix[cell][col], value);
        }

        var group = getGroup(this.matrix, row, col);

        for (cell = 0; cell < 9; cell++) {
            removeNumberFromCell(group[cell], value);
        }
    };

    this.checkMatrixForCraziness = function () {
        for (var i = 0; i < 9; i++) {
            for (var j = 0; j < 9; j++) {
                this.matrix[i][j].isCrazy = checkIfCellIsCrazy(this.matrix, i, j);
            }
        }
    };
};
